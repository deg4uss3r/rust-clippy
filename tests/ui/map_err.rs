#![warn(clippy::map_err_ignore)]
#![allow(clippy::unnecessary_wraps)]
use std::convert::TryFrom;
use std::error::Error;
use std::fmt;

#[derive(Debug)]
enum Errors {
    Ignored,
    OwnContext(u32),
    WithContext(std::num::TryFromIntError),
}

impl Error for Errors {}

impl fmt::Display for Errors {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Error")
    }
}

#[derive(Debug)]
struct SError {
    x: u32,
}

fn test_fn_call() -> u32 {
    0
}

fn main() -> Result<(), Errors> {
    let x = u32::try_from(-123_i32);
    let y = 0;

    // Should not warn you here, because you are giving context with a non-unit enum variant
    println!("{:?}", x.map_err(|_| Errors::OwnContext(0)));

    // Should not warn you here, because you are wrapping the context
    println!("{:?}", x.map_err(Errors::WithContext));

    // Should not warn you here, because you are giving context via a `String`
    println!("{:?}", x.map_err(|_| "There was an error!"));

    // Should not warn you here, because you are calling a constant for context
    println!("{:?}", x.map_err(|_| u32::MAX));

    // Should not warn you here, because you are calling a function for context
    println!("{:?}", x.map_err(|_| test_fn_call()));

    // Should not warn you here, because you are providing a variable for context
    println!("{:?}", x.map_err(|_| y));

    // Should not warn you here, because you are providing a struct for context
    println!("{:?}", x.map_err(|_| SError { x: 0 }));

    // Should warn you here because you are just ignoring the original error
    println!("{:?}", x.map_err(|_| Errors::Ignored));

    // Should not warn you because you explicitly ignore the parameter
    println!("{:?}", x.map_err(|_ignored_no_extra_context| Errors::Ignored));

    Ok(())
}
