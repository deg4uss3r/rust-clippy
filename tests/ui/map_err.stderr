error: `map_err(|_|...` ignores the original error
  --> $DIR/map_err.rs:57:32
   |
LL |     println!("{:?}", x.map_err(|_| Errors::Ignored));
   |                                ^^^
   |
   = note: `-D clippy::map-err-ignore` implied by `-D warnings`
   = help: Consider wrapping the error in an enum variant, or giving the closure parameter a name to intentionally ignore the error (e.g. `.map_err(|_ignored| ...)`

error: aborting due to previous error

